### PR TITLE
[5.7] Fix(TestResponse): assertExactJson false positive for JSON lists (ie non-assoc arrays

### LINKED ISSUE #26513: assertExactJson false positive for JSON lists (ie non-assoc arrays)
- Laravel Version: 5.7.11
- PHP Version: 7.1.3

### Description:

The `assertExactJson` method sorts the JSON, both in the response and in the expected data before comparing. This is not the correct way of handling JSON data if the data contains a list. Eg:
```json
{"foo": [1,2]}
```
is not the same JSON as
```json
{"foo": [2,1]}
```
but `assertExactJson` behaves as though they are equivalent.

In other words: for JSON comparison, associative arrays should be compared ignoring order, non-associative arrays should be compared respecting order.

### Steps To Reproduce:

```php
    public function testJson1()
    {
        // I would expect this to pass
        $response = new TestResponse((new JsonResponse(['foo' => [1, 2]])));
        $response->assertExactJson(['foo' => [1, 2]]);
    }

    public function testJson2()
    {
        // I would expect this to fail
        $response = new TestResponse((new JsonResponse(['foo' => [2, 1]])));
        $response->assertExactJson(['foo' => [1, 2]]);
    }
```

### PR DESCRIPTION
**Pull request by alberthuang24**
This PR fix #26513.

### CODE REVIEW

**Comment by vlakoff** [comment]
I think this method would be better named `assertSameJson`, as it doesn't check for exactly same JSON structure; it checks for same data, which can be presented differently (keyed values in different order).

**Comment by vlakoff** [comment]
Another thing, I don't like much `Arr::sortRecursive` because it sorts indexed arrays by values, and associative arrays by keys... This is quite a specific behaviour, and was designed specifically for this `assertExactJson` method.

Same goes for this new `Arr::sortAssocRecursive` method.

I think we may consider:
* deprecate/remove `Arr::sortRecursive` (and `array_sort_recursive`) as they would now be unused by Laravel,
* and move `Arr::sortAssocRecursive` to a protected method `TestResponse::sortAssocRecursive`

If you agree on the principle, this could be posted on [laravel/ideas](https://github.com/laravel/ideas), or if there is a clear consensus, just directly submit a PR.

**Comment by vlakoff** [comment]
If the above suggestion is agreed on, in this PR the new method `sortAssocRecursive` should be right now rather implemented as a protected method `TestResponse::sortAssocRecursive`, this way we wouldn't have to move it later.

File: src/Illuminate/Support/Arr.php [hunk 1]
```diff
@@ -594,6 +594,26 @@ public static function sortRecursive($array)
         return $array;
     }
 
+    /**
+     * Recursively sort an assoc array by keys.
+     * @param $array
```
Comment (vlakoff) [inline src/Illuminate/Support/Arr.php]: missing an empty line between description and params

File: src/Illuminate/Support/Arr.php [hunk 2]
```diff
@@ -596,7 +596,7 @@ public static function sortRecursive($array)
 
     /**
      * Recursively sort an assoc array by keys.
-     * @param $array
+     * @param  array  $array
```
Comment (vlakoff) [inline src/Illuminate/Support/Arr.php]: Yeah this too, but I meant the empty line between the description and the params :)

**Comment by driesvints** [comment]
@HongjiangHuang fwiw: it's best to explain in your main comment what you're trying to fix instead of linking to an issue.

**Comment by driesvints** [comment]
@HongjiangHuang can you also add a test for the `assertExactJson` method?

**Comment by taylorotwell** [comment]
No plans on changing the behavior of this function any further.

**Pull request closed without merging**